### PR TITLE
Fix the module of the hash function of a motion

### PR DESCRIPTION
**Pull request by gfphoenix78**
Fix the module of the hash function of a motion

After some refactors of the path/planner, the code assumes that
the number of receivers of a motion equals to the Gang size of
the parent slice. However it isn't always true, if some tables
are distributed on part of the segments. It could happen
during gpexpand.
Assume the following case, there is a GPDB cluster with 2 segments,
and the cluster is running `gpexpand` to add 1 segment. Now t1
and t2 are distributed on the first 2 segments, t3 has finished
data transfer, i.e. t3 is distributed on three segments.
See the following plan:
```
gpadmin=# explain select t1.a, t2.b from t1 join t2 on t1.a = t2.b union all select * from t3;
                                                QUERY PLAN
-----------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=2037.25..1230798.15 rows=7499310 width=8)
   ->  Append  (cost=2037.25..1080811.95 rows=2499770 width=8)
         ->  Hash Join  (cost=2037.25..1005718.85 rows=2471070 width=8)
               Hash Cond: (t2.b = t1.a)
               ->  Redistribute Motion 2:3  (slice2; segments: 2)  (cost=0.00..2683.00 rows=43050 width=4)
                     Hash Key: t2.b
                     ->  Seq Scan on t2  (cost=0.00..961.00 rows=43050 width=4)
               ->  Hash  (cost=961.00..961.00 rows=28700 width=4)
                     ->  Seq Scan on t1  (cost=0.00..961.00 rows=28700 width=4)
         ->  Seq Scan on t3  (cost=0.00..961.00 rows=28700 width=8)
 Optimizer: Postgres query optimizer
```
The slice2 shows that t2 will redistribute to all 3 segments and join
with t1. Since t1 is distributed only on the first 2 segments, the data
from t2 redistributed to the third segment couldn't have a match,
which returns wrong results.

The root cause is that the module of the cdb hash to redistribute t2 is
3, i.e. the Gang size of the parent slice. To fix this issue,
we add a field in Motion to record the number of receivers.
With this patch, the plan generated is:
```
gpadmin=# explain select * from t1  join t2 on t1.a = t2.b union all select a,a,b,b from t3;
                                                QUERY PLAN
----------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=2.26..155.16 rows=20 width=16)
   ->  Append  (cost=2.26..155.16 rows=7 width=16)
         ->  Hash Join  (cost=2.26..151.97 rows=7 width=16)
               Hash Cond: (t1.a = t2.b)
               ->  Seq Scan on t1  (cost=0.00..112.06 rows=5003 width=8)
               ->  Hash  (cost=2.18..2.18 rows=2 width=8)
                     ->  Redistribute Motion 2:2  (slice1; segments: 2)  (cost=0.00..2.18 rows=3 width=8)
                           Hash Key: t2.b
                           ->  Seq Scan on t2  (cost=0.00..2.06 rows=3 width=8)
         ->  Seq Scan on t3  (cost=0.00..3.06 rows=2 width=16)
 Optimizer: Postgres query optimizer
```
Note: the interconnect for the redistribute motion is still 2:3,
but the data transfer only happens in 2:2.

Co-authored-by: Zhenghua Lyu zlv@pivotal.io

- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/optimizer/plan/createplan.c [hunk 1]
```diff
@@ -7373,13 +7372,9 @@ cdbpathtoplan_create_motion_plan(PlannerInfo *root,
 {
 	Motion	   *motion = NULL;
 	Path	   *subpath = path->subpath;
+	/* numsegments is only used for hash motion */
```
Comment (kainwen) [inline src/backend/optimizer/plan/createplan.c:7375]: Change the name redistribute motion here.

And add some comments to explain why it is only used here.

File: src/backend/executor/nodeMotion.c [hunk 2]
```diff
@@ -766,6 +766,7 @@ ExecInitMotion(Motion *node, EState *estate, int eflags)
 	{
 		int			nkeys;
 
+		Assert(node->recv_numsegments >= 0 && node->recv_numsegments <= recvSlice->planNumSegments);
```
Comment (kainwen) [inline src/backend/executor/nodeMotion.c:769]: AssertImly here for orca?

use plangen.

File: src/backend/optimizer/plan/createplan.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/optimizer/plan/createplan.c]: renamed to numHashSegments

File: src/backend/executor/nodeMotion.c (hunk 2 above)
Comment (gfphoenix78) [reply src/backend/executor/nodeMotion.c]: Added.

File: src/test/regress/expected/union_gp.out [hunk 3]
```diff
@@ -1991,6 +1991,193 @@ select from t2_ncols union select * from t2_ncols;
...
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
+HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+create table union_schema.t3(a int, b int);
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
+HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+set allow_system_table_mods = on;
+update gp_distribution_policy set numsegments = 1
+  where localoid = 'union_schema.t1'::regclass::oid;
+update gp_distribution_policy set numsegments = 2
+  where localoid = 'union_schema.t2'::regclass::oid;
```
Comment (d) [inline src/test/regress/expected/union_gp.out:2013]:
```suggestion
update gp_distribution_policy set numsegments = 1
  where localoid = 'union_schema.t1'::regclass;
update gp_distribution_policy set numsegments = 2
  where localoid = 'union_schema.t2'::regclass;
```

Comment (gfphoenix78) [reply src/test/regress/expected/union_gp.out]: Thanks, updated.

**Comment by d** [comment]
Ha! This is a good problem to solve: one slice containing relations of different `numsegments` . We seem to have taken care to always enforce _some_ sort of uniformity in `JOIN`-like operations, like `INTERSECT` and `EXCEPT ALL`, so the only scenario I can think of is `UNION ALL`, is that right?

**Comment by d** [comment]
I played with this patch a little bit, very dumb question: it seems planner has a strong tendency to broadcast the inner side of an anti-join, regardless of its cardinality (example to follow). Is that caused by this patch, or does it already happen on master?

```sql
CREATE TEMP TABLE foo (a int, b int);
CREATE TEMP TABLE bar (c int, d int);
CREATE TEMP TABLE jazz (e int, f int);

BEGIN;
    SET LOCAL allow_system_table_mods TO on;
    UPDATE gp_distribution_policy SET numsegments = 1
        WHERE localoid = 'foo'::regclass;
    UPDATE gp_distribution_policy SET numsegments = 2
        WHERE localoid = 'bar'::regclass;
END;

SELECT localoid::regclass, policytype
FROM gp_distribution_policy
WHERE localoid in ('foo'::regclass, 'bar'::regclass)
ORDER BY 1;

INSERT INTO foo SELECT i, i FROM generate_series(3, 7) i;
INSERT INTO bar SELECT i, i FROM generate_series(1, 6) i;
INSERT INTO jazz SELECT i, i FROM generate_series(4, 1000003) i; -- doesn't seem to matter whether jazz has 1 tuple or 1 million
ANALYZE foo;
ANALYZE bar;
ANALYZE jazz;


SELECT $query$
SELECT c, a
FROM bar JOIN foo ON c = a
WHERE c NOT IN (SELECT e FROM jazz)
$query$ AS qry \gset

EXPLAIN
:qry;

:qry;

```

I got a plan like this:

```
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)  (cost=66103.21..66105.98 rows=20 width=8)
   ->  Hash Left Anti Semi (Not-In) Join  (cost=66103.21..66105.58 rows=10 width=8)
         Hash Cond: (bar.c = jazz.e)
         ->  Hash Join  (cost=1.21..3.34 rows=3 width=8)
               Hash Cond: (bar.c = foo.a)
               ->  Seq Scan on bar  (cost=0.00..2.06 rows=3 width=4)
               ->  Hash  (cost=1.15..1.15 rows=3 width=4)
                     ->  Redistribute Motion 1:2  (slice2; segments: 1)  (cost=0.00..1.15 rows=5 width=4)
                           Hash Key: foo.a
                           ->  Seq Scan on foo  (cost=0.00..1.05 rows=5 width=4)
         ->  Hash  (cost=41102.00..41102.00 rows=1000000 width=4)
               ->  Broadcast Motion 3:2  (slice3; segments: 3)  (cost=0.00..41102.00 rows=666667 width=4)
                     ->  Seq Scan on jazz  (cost=0.00..11102.00 rows=333334 width=4)
 Optimizer: Postgres query optimizer
(14 rows)

 c | a 
---+---
 3 | 3
(1 row)

```

**Comment by gfphoenix78** [comment]
@d No, this PR doesn't change how the plan is generated. It just records `numHashSegments` in the hash motion and it takes effect in the execution stage.

**Comment by gfphoenix78** [comment]
This could be in another story.

**Comment by d** [comment]
I don't think it's ideal we had to special-case `numHashSegments == 0` because we are afraid to change the ORCA translator code. For a little context, ORCA currently _only_ handles cases where every relation's `numsegments` equals the full cluster size. That means it should be a tractable problem to properly initialize this new field `Motion::numHashSegments` throughout the translator code. Do you need any help from @vraghavan78's team?

**Comment by gfphoenix78** [comment]
> I don't think it's ideal we had to special-case `numHashSegments == 0` because we are afraid to change the ORCA translator code. For a little context, ORCA currently _only_ handles cases where every relation's `numsegments` equals the full cluster size. That means it should be a tractable problem to properly initialize this new field `Motion::numHashSegments` throughout the translator code. Do you need any help from @vraghavan78's team?

Yes, I agree. It's better to set `numHashSegments` correctly when creating the motion. Is @vraghavan78  from ORCA team? It will be helpful. Thank you.

**Comment by d** [comment]
> Do you need any help from @vraghavan78's team?

Not sure I was clear the first time around: is there a specific reason you didn't touch up the code in `src/backend/gpopt` to set `Motion::numHashSegments` to a proper value? Do you want to ask somebody for help in knowing what to change?

**Comment by gfphoenix78** [comment]
> > Do you need any help from @vraghavan78's team?
> 
> Not sure I was clear the first time around: is there a specific reason you didn't touch up the code in `src/backend/gpopt` to set `Motion::numHashSegments` to a proper value? Do you want to ask somebody for help in knowing what to change?

I may find the place to set numHashSegments. @d  Thank you for your good suggestion.

**Comment by gfphoenix78** [comment]
Follow @d 's suggestion, dev pipeline still shows some failures. I'm fixing.

**Comment by gfphoenix78** [comment]
Make the dev pipeline green again.

**Pull request merged**